tidb-binlog/pump_client/selector.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"hash/fnv"
+	"strconv"
+	"sync"
+
+	pb "github.com/pingcap/tipb/go-binlog"
+)
+
+const (
+	// Hash means hash algorithm
+	Hash = "hash"
+
+	// Score means choose pump by it's score.
+	Score = "score"
+)
+
+// PumpSelector selects pump for sending binlog.
+type PumpSelector interface {
+	// SetPumps set pumps to be selected.
+	SetPumps(map[string]*PumpStatus)
+
+	// Select returns a situable pump.
+	Select(*pb.Binlog) *PumpStatus
+
+	// returns the next pump.
+	Next(*PumpStatus, *pb.Binlog, int) *PumpStatus
+}
+
+// HashSelector select a pump by hash.
+type HashSelector struct {
+	sync.RWMutex
+
+	// TsMap saves the map of start_ts with pump when send prepare binlog.
+	// And Commit binlog should send to the same pump.
+	TsMap map[int64]*PumpStatus
+
+	// PumpMap saves the map of pump's node id with pump.
+	PumpMap map[string]*PumpStatus
+
+	// the pumps to be selected.
+	Pumps []*PumpStatus
+}
+
+// NewHashSelector returns a new HashSelector.
+func NewHashSelector() PumpSelector {
+	return &HashSelector{
+		TsMap:   make(map[int64]*PumpStatus),
+		PumpMap: make(map[string]*PumpStatus),
+		Pumps:   make([]*PumpStatus, 0, 10),
+	}
+}
+
+// SetPumps implement PumpSelector.SetPumps.
+func (h *HashSelector) SetPumps(pumps map[string]*PumpStatus) {
+	h.Lock()
```

Comment:
put it before L73

Reply:
and argument `pumps` is not thread safe

Reply:
seems no need to chang lock place. 
I have already add lock before call this function.

Reply:
Do you think that makes sense? You've crossed two objects of the same rank with one lock